File: frontend/python-tools/mcp_tools/products/update_status.py
"""
Native MCP implementation for update_status
"""

from typing import Dict, Any, Optional
import sys
import os

# Add parent directory to path so we can import the original tools
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))

from base import ShopifyClient
from ..base import BaseMCPTool

class UpdateStatusTool(BaseMCPTool):
    """Update product status"""
    
    name = "update_status"
    description = "Change product status between active, draft, and archived"
    context = """
    Controls product visibility and availability:
    
    Status options:
    - ACTIVE: Product is visible and available for purchase
    - DRAFT: Product is hidden from storefront (for preparation/review)
    - ARCHIVED: Product is hidden and preserved for records
    
    Important notes:
    - Changing to draft/archived will hide product from customers
    - Archived products cannot be purchased even with direct link
    - Use draft for temporary hiding, archived for permanent removal
    - Status change affects ALL variants of a product
    """
    
    input_schema = {
        "type": "object", 
        "properties": {
            "product": {
                "type": "string",
                "description": "Product ID, handle, SKU, or title"
            },
            "status": {
                "type": "string",
                "enum": ["ACTIVE", "DRAFT", "ARCHIVED"],
                "description": "New product status"
            }
        },
        "required": ["product", "status"]
    }
    
    async def execute(self, product: str, status: str) -> Dict[str, Any]:
        """Update product status natively"""
        try:
            client = ShopifyClient()
            
            # Resolve product ID
            product_id = client.resolve_product_id(product)
            if not product_id:
                return {
                    "success": False,
                    "error": f"Product not found with identifier: {product}"
                }
            
            # Get current status first
            query = '''
            query getProductStatus($id: ID!) {
                product(id: $id) {
                    id
                    title
                    status
                    handle
                }
            }
            '''
            
            result = client.execute_graphql(query, {'id': product_id})
            product_data = result.get('data', {}).get('product')
            
            if not product_data:
                return {
                    "success": False,
                    "error": "Could not retrieve product"
                }
            
            current_status = product_data['status']
            
            # Check if status is already set
            if current_status == status:
                return {
                    "success": True,
                    "message": f"Product '{product_data['title']}' is already {status}",
                    "product": product_data
                }
            
            # Update status
            mutation = '''
            mutation updateProductStatus($input: ProductInput!) {
                productUpdate(input: $input) {
                    product {
                        id
                        title
                        status
                        handle
                    }
                    userErrors {
                        field
                        message
                    }
                }
            }
            '''
            
            variables = {
                'input': {
                    'id': product_id,
                    'status': status
                }
            }
            
            result = client.execute_graphql(mutation, variables)
            
            # Check for errors
            user_errors = result.get('data', {}).get('productUpdate', {}).get('userErrors', [])
            if user_errors:
                return {
                    "success": False,
                    "error": f"Update failed: {user_errors}"
                }
            
            updated_product = result.get('data', {}).get('productUpdate', {}).get('product')
            
            if updated_product:
                return {
                    "success": True,
                    "message": f"Successfully updated product status from {current_status} to {status}",
                    "product": {
                        "id": updated_product['id'],
                        "title": updated_product['title'],
                        "status": updated_product['status'],
                        "handle": updated_product['handle'],
                        "previous_status": current_status
                    }
                }
            else:
                return {
                    "success": False,
                    "error": "Update completed but product data not returned"
                }
                
        except Exception as e:
            return {
                "success": False,
                "error": str(e)
            }
            
    async def test(self) -> Dict[str, Any]:
        """Test the tool"""
        try:
            self.validate_env()
            client = ShopifyClient()
            
            # Test with a simple query to verify connection
            result = client.execute_graphql('{ shop { name } }')
            
            return {
                "status": "passed",
                "message": f"Connected to shop: {result.get('data', {}).get('shop', {}).get('name', 'Unknown')}"
            }
        except Exception as e:
            return {
                "status": "failed",
                "error": str(e)
            }

File: frontend/archive/legacy-nodejs-backend/server/native-tools/update-status.js
import { BaseTool } from './base-tool.js';

/**
 * Native JavaScript implementation of update_status tool
 * Updates product status (ACTIVE, DRAFT, ARCHIVED)
 */
export class UpdateStatusTool extends BaseTool {
  get metadata() {
    return {
      name: 'update_product_status',
      description: 'Update product status (ACTIVE, DRAFT, ARCHIVED)',
      inputSchema: {
        type: 'object',
        properties: {
          product_id: { 
            type: 'string', 
            description: 'Product ID (numeric or GID format)' 
          },
          status: { 
            type: 'string', 
            enum: ['ACTIVE', 'DRAFT', 'ARCHIVED'], 
            description: 'New product status' 
          }
        },
        required: ['product_id', 'status']
      }
    };
  }

  /**
   * Execute status update
   */
  async execute(args) {
    const { product_id, status } = args;
    
    this.log('Updating product status', { product_id, status });
    
    // Normalize product ID
    const productId = this.shopifyClient.normalizeId(product_id);
    
    // Validate status
    const validStatuses = ['ACTIVE', 'DRAFT', 'ARCHIVED'];
    if (!validStatuses.includes(status)) {
      throw new Error(`Invalid status: ${status}. Must be one of: ${validStatuses.join(', ')}`);
    }
    
    // Get current product details
    const query = `
      query getProductStatus($id: ID!) {
        product(id: $id) {
          id
          title
          status
        }
      }
    `;
    
    const data = await this.shopifyClient.query(query, { id: productId });
    
    if (!data.product) {
      throw new Error(`Product not found: ${productId}`);
    }
    
    const previousStatus = data.product.status;
    
    // Skip if status is already the same
    if (previousStatus === status) {
      return {
        product: data.product,
        message: `Product already has status: ${status}`,
        changed: false
      };
    }
    
    // Update product status
    const mutation = `
      mutation updateProductStatus($input: ProductInput!) {
        productUpdate(input: $input) {
          product {
            id
            title
            status
            publishedAt
            updatedAt
          }
          userErrors {
            field
            message
          }
        }
      }
    `;
    
    const variables = {
      input: {
        id: productId,
        status: status
      }
    };
    
    const result = await this.shopifyClient.mutate(mutation, variables);
    
    return {
      product: result.productUpdate.product,
      changes: {
        previousStatus,
        currentStatus: result.productUpdate.product.status
      },
      changed: true,
      message: `Product status updated from ${previousStatus} to ${status}`
    };
  }
}

// Export singleton instance
export const updateStatusTool = new UpdateStatusTool();

/**
 * Convenience tool for publishing products (setting status to ACTIVE)
 */
export class PublishProductTool extends BaseTool {
  get metadata() {
    return {
      name: 'publish_product',
      description: 'Publish a product by setting its status to ACTIVE',
      inputSchema: {
        type: 'object',
        properties: {
          product_id: { 
            type: 'string', 
            description: 'Product ID to publish' 
          }
        },
        required: ['product_id']
      }
    };
  }

  async execute(args) {
    const updateStatus = new UpdateStatusTool();
    return updateStatus.execute({
      product_id: args.product_id,
      status: 'ACTIVE'
    });
  }
}

export const publishProductTool = new PublishProductTool();

/**
 * Convenience tool for unpublishing products (setting status to DRAFT)
 */
export class UnpublishProductTool extends BaseTool {
  get metadata() {
    return {
      name: 'unpublish_product',
      description: 'Unpublish a product by setting its status to DRAFT',
      inputSchema: {
        type: 'object',
        properties: {
          product_id: { 
            type: 'string', 
            description: 'Product ID to unpublish' 
          }
        },
        required: ['product_id']
      }
    };
  }

  async execute(args) {
    const updateStatus = new UpdateStatusTool();
    return updateStatus.execute({
      product_id: args.product_id,
      status: 'DRAFT'
    });
  }
}

export const unpublishProductTool = new UnpublishProductTool();

/**
 * Convenience tool for archiving products
 */
export class ArchiveProductTool extends BaseTool {
  get metadata() {
    return {
      name: 'archive_product',
      description: 'Archive a product by setting its status to ARCHIVED',
      inputSchema: {
        type: 'object',
        properties: {
          product_id: { 
            type: 'string', 
            description: 'Product ID to archive' 
          }
        },
        required: ['product_id']
      }
    };
  }

  async execute(args) {
    const updateStatus = new UpdateStatusTool();
    return updateStatus.execute({
      product_id: args.product_id,
      status: 'ARCHIVED'
    });
  }
}

export const archiveProductTool = new ArchiveProductTool();